test/jdk/java/io/ObjectInputStream/NegativeArraySizeTest.java

```diff
@@ -0,0 +1,147 @@
+import java.io.InvalidClassException;
+import java.io.IOException;
+import java.io.ObjectInputFilter;
+import java.io.ObjectInputFilter.Status;
+import java.util.PriorityQueue;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.ObjectStreamException;
+
+public class NegativeArraySizeTest {
+
+    private static byte[] buildArrayPayload() throws IOException {
+         // Serialize to bytes
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        ObjectOutputStream oos = new ObjectOutputStream(baos);
+        oos.writeObject(new String[1]);
+        oos.close();
+        byte[] serializedData = baos.toByteArray();
+
+        // Find the right location to modify, looking for the first instance of TC_ENDBLOCKDATA
+        int firstPos = 0;
+        for (int i = 0; i < serializedData.length - 1; i++) {
+            // 0x78 = TC_ENDBLOCKDATA
+            if (serializedData[i] == 0x78) {
+                // Replace array length with -2
+                serializedData[i + 2] = (byte) 0xff;
+                serializedData[i + 3] = (byte) 0xff;
+                serializedData[i + 4] = (byte) 0xff;
+                serializedData[i + 5] = (byte) 0xfe;
+
+                return serializedData;
+            }
+        }
+        throw new RuntimeException("Can't find TC_ENDBLOCKDATA in object output stream");
+    }
+
+    private static byte[] buildPriorityQueuePayload() throws IOException {
+        // Serialize to bytes
+       ByteArrayOutputStream baos = new ByteArrayOutputStream();
+       ObjectOutputStream oos = new ObjectOutputStream(baos);
+       oos.writeObject(new PriorityQueue<>());
+       oos.close();
+       byte[] serializedData = baos.toByteArray();
+
+       // Find the right location to modify, looking for the first instance of TC_ENDBLOCKDATA
+       int firstPos = 0;
+       for (int i = 0; i < serializedData.length - 1; i++) {
+           // 0x77 = TC_BLOCKDATA
+           if (serializedData[i] == 0x77) {
+               // Replace array length with -3
+               serializedData[i - 5] = (byte) 0xff;
+               serializedData[i - 4] = (byte) 0xff;
+               serializedData[i - 3] = (byte) 0xff;
+               serializedData[i - 2] = (byte) 0xfd;
+
+               return serializedData;
+           }
+       }
+       throw new RuntimeException("Can't find TC_BLOCKDATA in object output stream");
+    }
+
+    private static class CustomFilter implements ObjectInputFilter {
+        @Override
+        public Status checkInput(FilterInfo filterInfo) {
+            Class<?> cl = filterInfo.serialClass();
+            if (cl != null && cl.isArray() && filterInfo.arrayLength() < -1) {
+                throw new RuntimeException("FilterInfo.arrayLength() must be >= -1 for arrays (was " + filterInfo.arrayLength() + ")");
+            }
+            return Status.ALLOWED;
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        try {
```

Comment:
Test code nits, we can probably `try-with-resources` this thing to make it cleaner and make sure nothing unexpectedly bad is happening on "close" paths.

Reply:
OK, here we go...